File: src/java/src/org/genboree/util/TimingUtil.java
package org.genboree.util;

import org.genboree.util.Util;
import java.io.*;
import java.sql.*;
import java.util.*;

public class TimingUtil
{
  // OBJECT VARIABLES
  protected ArrayList vTimes ;
  protected ArrayList vMsgs ;
  protected String[] userInfo ;
  protected boolean forced ;
  
  public TimingUtil()
  {
    this(null) ;
    this.forced = true ;
  }
  
  public TimingUtil(String[] userInfo)
  {
    this.vTimes = new ArrayList() ;
    this.vMsgs = new ArrayList() ;
    this.userInfo = userInfo ;
    this.forced = false ;
    
    java.util.Date now = new java.util.Date() ;
    this.vTimes.add( now );
    this.vMsgs.add( "TIMING STARTED AT: " + now.toString() ) ;
  }
  
  public void addMsg(String msg)
  {
    this.vTimes.add(new java.util.Date()) ;
    this.vMsgs.add(msg) ;
    return ;
  }
  
  public void writeTimingReport(PrintStream ios)
  {
    this.writeTimingReport(new PrintWriter(ios)) ;
    return ;
  }
  
  public void writeTimingReport(Writer writer)
  {
    PrintWriter ios = new PrintWriter(writer) ;

    if(this.forced || (this.userInfo != null) || this.userInfo[0].equals("andrewj"))
    {
      java.util.Date currDate = (java.util.Date) this.vTimes.get(0) ;
      String currMsg = (String) this.vMsgs.get(0) ;
      ios.println( currMsg ) ;
      ios.println();

      java.util.Date nextDate = null ;
      for( int ii=1; ii<this.vTimes.size(); ii++ )
      {
        nextDate = (java.util.Date) this.vTimes.get(ii) ;
        currMsg = (String) this.vMsgs.get(ii) ;
        long elapsedTime = (nextDate.getTime() - currDate.getTime()) / 100L ;
        ios.println( "    " + currMsg + ": " + (elapsedTime/10L) + "." + (elapsedTime % 10) + " sec" );
        currDate = nextDate ;
      }

      ios.println() ;
      ios.println( "TIMING REPORTED AT: " + nextDate.toString() );
    }
    ios.flush() ;
    return ;
  }

    public String generateStringWithReport()
    {
        StringBuffer report = new StringBuffer(100);

        if(this.forced || (this.userInfo != null && (this.userInfo[0].equals("andrewj"))))
        {
            java.util.Date currDate = (java.util.Date) this.vTimes.get(0) ;
            String currMsg = (String) this.vMsgs.get(0) ;
            report.append(currMsg).append("\n");

            java.util.Date nextDate = null ;
            for( int ii=1; ii<this.vTimes.size(); ii++ )
            {
                nextDate = (java.util.Date) this.vTimes.get(ii) ;
                currMsg = (String) this.vMsgs.get(ii) ;
                long elapsedTime = (nextDate.getTime() - currDate.getTime()) / 100L ;
                report.append("    ").append(currMsg).append(": ");
                report.append(elapsedTime/10L).append(".");
                report.append(elapsedTime % 10).append(" sec").append("\n");
                currDate = nextDate ;
            }
//            report.append("\n");
            report.append("TIMING REPORTED AT: ").append(nextDate.toString()).append("\n");
        }
        return report.toString();

    }

}

File: src/ruby/brl/util/timingUtil.rb
#!/usr/bin/env ruby

# ##############################################################################
# REQUIRED CLASSES/MODULES
# ##############################################################################


module BRL ;  module Util

class TimingUtil
  # ############################################################################
	# CONSTANTS
	# ############################################################################
	
	# ############################################################################
	# ATTRIBUTES
	# ############################################################################
  attr_accessor :vTimes, :vMsgs
  
  def initialize()
    @vTimes = []
    @vMsgs = []
    self.addMsg("TIMING STARTED AT: " + Time.now().to_s)
  end
  
  def addMsg(msg)
    @vTimes << Time.now()
    @vMsgs << msg 
    return
  end
    
  def writeTimingReport(writer)
    currDate = @vTimes[0]
    currMsg = @vMsgs[0]
    writer.puts currMsg 
    writer.puts '' 

    nextDate = nil
    (1...@vTimes.size()).each { |ii|
      nextDate = @vTimes[ii]
      currMsg = @vMsgs[ii]
      elapsedTime = (nextDate - currDate)
      writer.puts "    #{currMsg}: #{elapsedTime.to_i / 60} min #{elapsedTime.to_i % 60} sec"
      currDate = nextDate
    }
    writer.puts ''
    writer.puts "TIMING REPORTED AT: " + nextDate.to_s
    writer.flush() if(writer.respond_to?(:flush))
    return
  end
  
  alias :<< :addMsg
end

end ; end

# ############################################################################
# MAIN (quick unit test)
# ############################################################################
if(__FILE__ == $0)
  @timer = BRL::Util::TimingUtil.new()
  puts "\n\nStart unit test; 20 sec sleep with addMsg() test..."
  sleep(20)
  @timer.addMsg("- Done sleeping for 20 sec")
  puts "...done 20 sec sleep. Now do 65 sec sleep with << test..."
  sleep(65)
  @timer << "- Done sleeping for 65 more sec"
  puts "...done 65 sec sleep. Now write timing report to $stdout and $stderr."
  @timer.writeTimingReport($stdout)
  @timer.writeTimingReport($stderr)
end
